compiler/optimizations/deadCodeElimination.cpp

```diff
@@ -434,6 +438,248 @@ static void deadModuleElimination() {
   }
 }
 
+static bool shouldOutlineLoop(CForLoop *loop) {
+  // Obvious TODO :)
+  
+  const char* testModuleName = "GPUOutlineTest";
+  bool inTestModule = false;
+
+  Symbol* cur = loop->parentSymbol;
+
+  do {
+    if (ModuleSymbol* parentModule = toModuleSymbol(cur)) {
+      if (strcmp(parentModule->name, testModuleName) == 0) {
+        inTestModule = true;
+      }
+    }
+
+    if (cur->defPoint != NULL)
+      cur = cur->defPoint->parentSymbol;
+    else
+      cur = NULL;
+
+  } while (cur != NULL && !inTestModule);
+
+  return inTestModule && loop->isOrderIndependent();
+}
+
+static bool isDefinedInTheLoop(Symbol* sym, CForLoop* loop) {
+  return LoopStmt::findEnclosingLoop(sym->defPoint) == loop;
+}
+
+static bool isIndexVariable(Symbol* sym, CForLoop* loop) {
+
+  // This is really a poor check. My hope is that we can achieve this here much
+  // simpler by associating index variable(s) with CForLoops when we create
+  // them. For now, this prototype only checks for an `idx = ` in the initBlock
+  // of the C loop.
+  std::vector<CallExpr*> calls;
+  collectCallExprs(loop->initBlockGet(), calls);
+
+  for_vector (CallExpr, call, calls) {
+    if (call->isPrimitive(PRIM_ASSIGN)) {
+      SymExpr* lhsSE = toSymExpr(call->get(1));
+      INT_ASSERT(lhsSE);
+
+      if (lhsSE->symbol() == sym) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+static  CallExpr* generateGPUCall(FnSymbol* kernel,
+                                     std::vector<VarSymbol*> actuals) {
+  CallExpr* call = new CallExpr(PRIM_GPU_KERNEL_LAUNCH);
+  call->insertAtTail(new_CStringSymbol(kernel->cname));
+
+  call->insertAtTail(new_IntSymbol(1));  // grid size
+  call->insertAtTail(new_IntSymbol(1));  // block size
+
+  call->insertAtTail(new_IntSymbol(actuals.size()));
+
+  for_vector (VarSymbol, actual, actuals) {
+    call->insertAtTail(new SymExpr(actual));
+  }
+
+  return call;
+}
+
+
+static void outlineGPUKernels() {
+  forv_Vec(FnSymbol*, fn, gFnSymbols) {
+    std::vector<BaseAST*> asts;
+    collect_asts(fn, asts);
+
+    for_vector(BaseAST, ast, asts) {
+      if (CForLoop* loop = toCForLoop(ast)) {
+        if (shouldOutlineLoop(loop)) {
+          SET_LINENO(loop);
+          FnSymbol* outlinedFunction = new FnSymbol("chpl_gpu_kernel");
```

Comment:
> Does the Chapel compiler have some kind of renamer pass to uniqueify all the symbols? Or will we run into issues if we create multiple functions named chpl_gpu_kernel?

codegen adds numbers if there are more than one symbol with the same name. So, no need to worry about it here.

> I'm also curious if you think we would care at all about the name of the generated function since we're no intending this to be used for C code generation. If we do care (for debugging purposes or otherwise) maybe we could make the identifier incorporate the filename and line number it originated from.

That's interesting. Back in the C backend world, we would put a comment right before the function that says something like

```
/* MyTest.chpl: 100 */
```

Inspecting the generated LLVM is a bit more different, and as far as I know we don't have any marker that can help us with that. I think this is more about LLVM workflow rather than this PR, but I'll create a private issue to see if there are methods that I'm unaware of.

And as a general debugging pointer, the AST nodes have their location info, and you can use `loc` in the debugger to get that. `SET_LINENO` above is the macro magic that sets the line numbers of every AST created in this block.

Reply:
https://github.com/Cray/chapel-private/issues/2353